### PR TITLE
Mosaic memory profiling tutorial

### PR DESCRIPTION
**Pull request by basilwong**
## Description

This PR adds a comprehensive tutorial on using [Mosaic](https://github.com/facebookresearch/mosaic) for GPU memory profiling in PyTorch.

Mosaic is a post-analysis tool for memory usage that was instrumental in debugging OOM issues during the 405B LLaMA training.

## What users will learn

1. **Categorical Memory Profiling** - Breaking down memory by category (activation, gradient, optimizer, parameters)
2. **Debugging Unexpected Memory** - Using stack trace analysis to find abandoned debug code causing memory bloat
3. **Pipeline Integration** - Using Mosaic's Python API for automated memory monitoring and CI/CD regression testing

## Tutorial structure

- Introduction to Mosaic and installation
- Simple usage examples (CLI commands)
- Real-World Case 1: Activation Checkpointing Analysis
- Real-World Case 2: Debugging Unexpected Memory Usage  
- Real-World Case 3: Pipeline Integration with Python API

## Requirements

- PyTorch with CUDA support
- `pip install git+https://github.com/facebookresearch/mosaic.git`
- GPU required to run the examples

## Checklist

- [x] Tutorial runs without errors
- [x] Tutorial follows sphinx-gallery format
- [x] Images included in `_static/img/mosaic/`
- [ ] CI passes

## Related Links

- Mosaic Repository: https://github.com/facebookresearch/mosaic

### CODE REVIEW

File: beginner_source/mosaic_memory_profiling_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,1175 @@
+# -*- coding: utf-8 -*-
+
+"""
+Mosaic: Memory Profiling for PyTorch
+====================================
```
Comment (svekars) [inline beginner_source/mosaic_memory_profiling_tutorial.py:5]: Suggesting to add meta info for better SEO
```suggestion
====================================

.. meta::
   :description: Learn how to use Mosaic for PyTorch GPU memory profiling. Capture and analyze memory snapshots, identify memory savings from activation checkpointing, debug OOM errors, and integrate memory analysis into training pipelines.
   :keywords: PyTorch, Mosaic, memory profiling, GPU memory, CUDA, activation checkpointing, OOM debugging, deep learning, memory optimization, memory snapshots, distributed training, LLaMA, transformer models
```

File: beginner_source/mosaic_memory_profiling_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,1175 @@
+# -*- coding: utf-8 -*-
```
Comment (svekars) [inline beginner_source/mosaic_memory_profiling_tutorial.py:1]: Need to add to a toctree, probably in the Ecosystem category:  https://github.com/pytorch/tutorials/blob/main/ecosystem.rst?plain=1#L63 and add a .. customcarditem.

I also suggest to do at to the What's New in Pytorch Tutorials to increase visibility: https://github.com/pytorch/tutorials/blob/main/index.rst?plain=1#L4C1-L4C37 

Add a customcard in the index.rst as well.

Comment (basilwong) [reply beginner_source/mosaic_memory_profiling_tutorial.py:1]: Made changes based on your suggestions 👍
